### PR TITLE
Fix crash in Missle that tries to sqrt number < 0

### LINKED ISSUE #19390: Crash related with missile climbs higher terrain
## System Information

- Operating System:  widows 10 
- MissileTA.cs modified from Missile.cs from release version last year, but crash related place was not modified since then.

## Exception log

```
OpenRA engine version 05312d7
Tiberian Aurora mod version {DEV_VERSION}
on map 97ab1d41cd0e37f1024ce8cfb7da174daf0b27fd (Chaos Valley by 零现在是小猫咪).
Date: 2021-05-04 05:03:46Z
Operating System: Windows (Microsoft Windows NT 10.0.18363.0)
Runtime Version: .NET CLR 5.0.4
Exception of type `System.InvalidOperationException`: Attempted to calculate the square root of a negative integer: -1799301
   at OpenRA.Exts.ISqrt(Int32 number, ISqrtRoundMode round) in D:\TA-Main\TiberianAurora\engine\OpenRA.Game\Exts.cs:line 281
   at OpenRA.Mods.Common.Projectiles.MissileTA.HomingInnerTick(Int32 predClfDist, Int32 diffClfMslHgt, Int32 relTarHorDist, Int32 lastHtChg, Int32 lastHt, Int32 nxtRelTarHorDist, Int32 relTarHgt, Int32 vFacing, Boolean targetPassedBy) in D:\TA-Main\TiberianAurora\OpenRA.Mods.TA\Projectiles\MissileTA.cs:line 689
   at OpenRA.Mods.Common.Projectiles.MissileTA.HomingTick(World world, WVec tarDistVec, Int32 relTarHorDist) in D:\TA-Main\TiberianAurora\OpenRA.Mods.TA\Projectiles\MissileTA.cs:line 824
   at OpenRA.Mods.Common.Projectiles.MissileTA.Tick(World world) in D:\TA-Main\TiberianAurora\OpenRA.Mods.TA\Projectiles\MissileTA.cs:line 923
   at OpenRA.WorldUtils.DoTimed[T](IEnumerable`1 e, Action`1 a, String text) in D:\TA-Main\TiberianAurora\engine\OpenRA.Game\WorldUtils.cs:line 78
   at OpenRA.World.Tick() in D:\TA-Main\TiberianAurora\engine\OpenRA.Game\World.cs:line 441
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in D:\TA-Main\TiberianAurora\engine\OpenRA.Game\Game.cs:line 646
   at OpenRA.Game.LogicTick() in D:\TA-Main\TiberianAurora\engine\OpenRA.Game\Game.cs:line 661
   at OpenRA.Game.Loop() in D:\TA-Main\TiberianAurora\engine\OpenRA.Game\Game.cs:line 823
   at OpenRA.Game.Run() in D:\TA-Main\TiberianAurora\engine\OpenRA.Game\Game.cs:line 864
   at OpenRA.Game.InitializeAndRun(String[] args) in D:\TA-Main\TiberianAurora\engine\OpenRA.Game\Game.cs:line 274
   at OpenRA.Launcher.Program.Main(String[] args) in D:\TA-Main\TiberianAurora\engine\OpenRA.Launcher\Program.cs:line 30
```

## Additional information

The crash is due to `2 * loopRadius - h1` < 0, as you can see at picture below. 

I try get var data by using:

When crash happens, the var data here is:

Note:
As for somewhere in "MissileTA" may affect calculation, I allow this missile will not explode when its `DistanceAboveTerrain` < 0. Don't know if this will affect the height calculation. If this affects little, the crash will affects TS style map.

### PR DESCRIPTION
**Pull request by dnqbob**
Fixes #19390

### CODE REVIEW

**Comment by pchote** [comment]
Can you please explain the logic behind the fix here?
Its not clear from the diff why these heights can only now be greater than 2 * loopRadius or why setting these to 0 is the correct fix.

**Comment by dnqbob** [comment]
> Its not clear from the diff why these heights can only now be greater than 2 * loopRadius or why setting these to 0 is the correct fix.

As for "why these heights can only now be greater than 2 * loop radius", I don't know either on the original method they used. The fix I done here is only a <0 detection to avoid game crash on height < 2 * loop radius (although it is rare, it does crash) 

Better idea is welcomed and in needed.

File: OpenRA.Mods.Common/Projectiles/Missile.cs [hunk 1]
```diff
@@ -559,7 +559,7 @@ int IncreaseAltitude(int predClfDist, int diffClfMslHgt, int relTarHorDist, int
 
 					|| (desiredVFacing == 0 // Upper part of incline surmounting manoeuvre
 						&& relTarHorDist <= loopRadius * WAngle.FromFacing(vFacing).Sin() / 1024
-							+ Exts.ISqrt(predAttHght * (2 * loopRadius - predAttHght)))); // Target too close to hit at current speed
+							+ Exts.ISqrt(predAttHght * (2 * loopRadius - predAttHght > 0 ? 2 * loopRadius - predAttHght : 0)))); // Target too close to hit at current speed
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Projectiles/Missile.cs:562]: perhaps
```suggestion
							+ Exts.ISqrt(Math.Abs(predAttHght * (2 * loopRadius - predAttHght))))); // Target too close to hit at current speed
```

File: OpenRA.Mods.Common/Projectiles/Missile.cs [hunk 2]
```diff
@@ -659,7 +659,10 @@ int IncreaseAltitude(int predClfDist, int diffClfMslHgt, int relTarHorDist, int
 							var h1 = loopRadius - Exts.ISqrt(d1 * (2 * loopRadius - d1)) - (pos.Z - lastHt);
 
 							if (h1 > loopRadius * (1024 - WAngle.FromFacing(vFacing).Cos()) / 1024)
-								desiredVFacing = WAngle.ArcTan(Exts.ISqrt(h1 * (2 * loopRadius - h1)), loopRadius - h1).Angle >> 2;
+							{
+								var maxVChange = 2 * loopRadius - h1 > 0 ? 2 * loopRadius - h1 : 0;
+								desiredVFacing = WAngle.ArcTan(Exts.ISqrt(h1 * maxVChange), loopRadius - h1).Angle >> 2;
+							}
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Projectiles/Missile.cs:665]: and same here?

**Review by PunkPun — COMMENTED** [review]
To me it's unclear what the code is trying to do, but this sulution just seems wrong. To avoid the crash we could use `Math.Abs(variable)` or `Math.Max(variable, 0)`

Also please rebase this is 730 commits behind

File: OpenRA.Mods.Common/Projectiles/Missile.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Projectiles/Missile.cs:562]: idk if we should use an absolute number here, also I don't know how to produce an usecase to test this function properly(loop radius and other parameters are too substract). All I know is, iosmeric map with height will have a tiny chance crash on it.

Comment (PunkPun) [reply OpenRA.Mods.Common/Projectiles/Missile.cs:562]: as I mentioned we could use Math.Max(variable, 0) instead

Comment (dnqbob) [reply OpenRA.Mods.Common/Projectiles/Missile.cs:562]: Tomorrow I will take a look at it. I think I should make a comment here about the future missile test.

**Comment by dnqbob** [comment]
This branch is abandoned long time and missile also has other bugs on isometric map. I think rework missile can be a better way.

**Comment by PunkPun** [comment]
Sure, we still don't officially support isometric mods so it might be fine to wait for a proper fix

**Pull request closed without merging**